Use Proxy.revokable and add revokeProxy to API

alshakero opened the pull request:
I'm implementing the needed specs right now.

Implements: https://github.com/Palindrom/Palindrom/issues/134

warpech commented on src/jsonpatcherproxy.js:
```diff
@@ -8,6 +8,7 @@
 /** Class representing a JS Object observer  */
 var JSONPatcherProxy = (function() {
   function JSONPatcherProxy(root) {
+    this.proxifiedObjectsMap = [];
```
A WeakMap here would be more memory-friendly.

warpech commented on src/jsonpatcherproxy.js:
```diff
@@ -143,7 +144,9 @@ var JSONPatcherProxy = (function() {
+    return proxy.proxy;
   };
   //grab tree's leaves one by one, encapsulate them into a proxy and return
   JSONPatcherProxy.prototype._proxifyObjectTreeRecursively = function(
```
Could `_proxifyObjectTreeRecursively` automatically revoke proxy of when a child object is removed?

alshakero replied:
@warpech I started with this yesterday but figured that would render the child object useless even if it was cached earlier on purpose for later usages, [like in this demo](https://jsbin.com/barexugoca/1/edit?js,console). Doing it in Palindrom made more sense since `Palindrom.obj` is ours and we give it to the user for use. But the observed object in JSONPatcherProxies is not ours and I guess we should leave it, and its children, usable. 

What do you think?

alshakero replied:
I see my mistake here. It will emit patches to the same callback which is problematic. I think disabling the traps inside `_proxifyObjectTreeRecursively` would be better.

warpech replied:
Hmm... this is a philosophical question now :) I am not sure. Maybe let's see how it goes with controlling it from Palindrom.

warpech replied:
Great, then let's do it like you say.

tomalec replied:
@alshakero  The jsbin throws 504 to me :| could you send it once again?

Forgive me if I get something wrong. I still don't get why can we `revoke` on `deleteProperty` by default.

> But the observed object in JSONPatcherProxies is not ours and I guess we should leave it, and its children, usable.

If child **proxy** object "was cached earlier on purpose for later usages". It's a proxy object, so it's ours and we can destroy it. If user wants to cache his/hers object for later usages after it's unobserved by us, he/she can use `originalObject` or the object returned by delete/splice - I believe it's the original one https://github.com/Palindrom/JSONPatcherProxy/pull/4/files?w=1#diff-3bfa726203c6cb018f2321c18ca9298eR94 . I'm not sure, @alshakero is it?

alshakero replied:
@tomalec now I implemented the ability to disable traps for a given proxy, [and this happens automatically in `deleteProperty` trap](https://github.com/Palindrom/JSONPatcherProxy/pull/4/commits/b1022e8336e256f0f448770987ad12e12a07296f#diff-3bfa726203c6cb018f2321c18ca9298eR91). It's more fail-proof. And I regret the revoking proposal, it's redundant now. I'll explain:

If an element (say an HTML `input` field) is bound to as sub-object of the state tree using Polymer binding. And a patch comes to delete that sub-object from the server, editing the field will throw a TypeError (in the time window of Polymer binding async update), and this error will be to the user in production, not to the developer. Any race condition will cause this issue and it will be very hard to reproduce. 

WDYT?

alshakero replied:
@tomalec the whole JSBin is down 🤒

tomalec replied:
Could you use jsfiddle.net?

Once delete patch is applied binding should be removed as well.
If you are talking about race conditions between browser-to-server and server-to-browser messages, OT should solve it.

tomalec replied:
Could we setup a HO call, I feel I miss something in here.

warpech commented on src/jsonpatcherproxy.js:
```diff
@@ -177,6 +180,23 @@ var JSONPatcherProxy = (function() {
+   */
+  JSONPatcherProxy.prototype.revokeProxy = function(proxy) {
+      let deletedIndex;
+      this.proxifiedObjectsMap.forEach(function(el, index) {
```
With WeakMap this becomes simpler: `this.proxifiedObjectsMap.get(proxy)`.

alshakero replied:
Totally forgot about its whole existence! Great idea.

tomalec commented:
Just added a comment at https://github.com/Palindrom/JSONPatcherProxy/pull/4/files/a50598b02b1a7c4f34e89afcf5e902defeb83d52#r115702691 but GH hides it.

alshakero commented:
@tomalec here is my response in case it was hidden:  https://github.com/Palindrom/JSONPatcherProxy/pull/4#discussion_r115706017

alshakero commented:
@tomalec I think this is ready for review.

tomalec commented on src/jsonpatcherproxy.js:
```diff
@@ -132,18 +84,87 @@ var JSONPatcherProxy = (function() {
+      return obj;
+    }
+    var instance = this;
+    var trapsInstance = Object.assign({}, traps, {path: path, instance: instance});
```
isn't 
```js
trapsInstance.path = path;
trapsInstance.instance = instance;
```
cheaper?

Or even just define traps in here, without a need to create 3 new objects?

alshakero replied:
That's right. I thought not defining the functions (`set`, `get` etc..) again and just referencing them would be better. But this would be slower with every issued patch (`this.instance.defaultCallback` is slow).

Moving them inside could be a little slower but for one time only. Faster with every patch.

I moved it inside.

alshakero commented on README.md:
```diff
@@ -119,18 +124,21 @@ It returns the changes of your object since the last time it's called. You have
+#### JSONPatcherProxy#resume () : void
 
-#### JSONPatcherProxy#switchCallbackOff () : void
+Enables patches omitting (to both callback and patches array). Starting from the moment you call it.
```
What an unlucky typo. I mean emitting.

alshakero commented:
@tomalec is this good to merge?

tomalec commented on src/jsonpatcherproxy.js:
```diff
@@ -6,88 +6,98 @@
+  /**
+    * Creates an instance of JSONPatcherProxy around your object of interest `root`. 
+    * @param {Object|Array} root - the object you want to wrap
+    * @param {Boolean} showDetachedWarning - whether to log a warning when a detached sub-object is modified @see {@link https://github.com/Palindrom/JSONPatcherProxy#detached-objects} 
```
WDYT about adding optional parameter info and default values in jsdoc? http://usejsdoc.org/tags-param.html#optional-parameters-and-default-values

tomalec commented on src/jsonpatcherproxy.js:
```diff
@@ -196,28 +240,43 @@ var JSONPatcherProxy = (function() {
+   * Disables all proxies' traps, turning the observed object into a forward-proxy object, like a normal object that you can modify silently.
+   */
+  JSONPatcherProxy.prototype.disableTraps = function() {
+    proxifiedObjectsMap.forEach(disableTrapsForProxy.bind(this));
```
Why don't we make `disableTrapsForProxy` instance method as we are always calling it on instance - it results in creating new function of every call, here and on delete? Do you want to keep it private?

tomalec commented on test/spec/proxySpec.js:
```diff
@@ -993,167 +1021,280 @@ describe("proxy", function () {
 
-            observedObj.firstName = 'Malvin';            
-            expect(called).toReallyEqual(1);
+            it('should omit patches when unobserved then observed', function () {
```
We are testing `pause` and `resume`  not "unobservered" then "observed"

tomalec commented on test/spec/proxySpec.js:
```diff
@@ -993,167 +1021,280 @@ describe("proxy", function () {
+                observedObj.firstName = 'Carlos';       
+                expect(called).toReallyEqual(3);
             });
+            it('should unobserve then observe again (deep value)', function () {
```
and we have `unobserve` again, It's quite confusing to me to mix `pause` and `unobserve` test together, under `describe('unobserving', function(){`

tomalec commented:
Overall I think it's just cosmetics, but `unobserve` suite in test is quite confusing to me, wouldn't it be cleaner to have separate section for `pause`?

alshakero commented:
@tomalec we have done all kinds of breaking changes, why should we keep `unobserve` in its current utterly useless case. I removed it and stuck with `pause`, `resume`, and `disableTraps`. And cleaned up the tests.

tomalec commented on test/spec/proxySpec.js:
```diff
@@ -844,457 +943,421 @@ describe("proxy", function () {
+  });
+
+  describe('pausing and resuming', function() {
+    it("shouldn't omit patches when paused", function() {
```
I think it should be: "Should omit patches when paused"

tomalec commented on test/spec/proxySpec.js:
```diff
@@ -844,457 +943,421 @@ describe("proxy", function () {
+      expect(called).toReallyEqual(2);
+    });
+
+    it('should omit patches when paused then resumed', function() {
```
I'd say: "Should re-start emitting  patches when paused then resumed"

alshakero commented:
Good to go?
